src/plugins/pcr_primer_design_for_dna_assembly/src/options_panel/UserPimerLineEdit.cpp

```diff
@@ -0,0 +1,216 @@
+        }
+        if (rightStartPos <= endSelectionPos && endSelectionPos <= rightEndPos) {
+            rightEndSize = rightEndPos - endSelectionPos;
+        }
+    }
+}
+
+void UserPimerLineEdit::sl_textChanged(const QString& text) {
+    if (lastSelectedTextSize == 0 && !changeLeftEndColor && !changeRightEndColor) {
+        // We don't have a selection, so it's either
+        // an insertion or deletion, but not both.
+        int delta = text.size() - lastTextSize;
+        if (delta > 0) {
+            // User has inserted text.
+            int pos = cursorPosition() - delta;
+            for (int i = 0; i < delta; i++) {
+                colors.insert(pos, getColorForInsertedText(pos));
+            }
+        } else {
+            // User has erased text.
+            int pos = cursorPosition();
+            colors.erase(colors.begin() + pos, colors.begin() + pos - delta);
+        }
+    } else {
+        // There was a selection, so we have both removed
+        // and inserted text.
+        int pos = lastSelectionStart;
+        int removedCount = lastSelectedTextSize;
+        int insertedCount = cursorPosition() - pos;
+        if (changeLeftEndColor) {
+            pos = 0;
+            removedCount = lastLeftEndSize;
+            insertedCount = leftEndSize;
+        } else if (changeRightEndColor) {
+            pos = leftEndSize + qMax(0, lastTextSize - (leftEndSize + lastRightEndSize));
+            removedCount = lastRightEndSize;
+            insertedCount = rightEndSize;
+        }
+
+        colors.erase(colors.begin() + pos, colors.begin() + pos + removedCount);
+        for (int i = 0; i < insertedCount; i++) {
+            colors.insert(pos, getColorForInsertedText(pos));
+        }
+    }
+    changeLeftEndColor = false;
+    changeRightEndColor = false;
+
+    setCharColors(colors);
+}
+
+void UserPimerLineEdit::setCharColors(const QList<QColor>& newColors) {
+    QList<QInputMethodEvent::Attribute> attributes;
+    int size = newColors.size();
+    attributes.reserve(size);
+    for (int i = 0; i < size; i++) {
+        if (newColors[i].isValid()) {
+            QTextCharFormat charFormat;
+            charFormat.setForeground(QBrush(newColors[i]));
+            const int start = i - cursorPosition();
+            const int length = 1;
+            attributes.append(QInputMethodEvent::Attribute(QInputMethodEvent::TextFormat, start, length, charFormat));
+        }
+    }
+    QInputMethodEvent ime(QString(), attributes);
+    QLineEdit::inputMethodEvent(&ime);
+
+    lastTextSize = text().size();
+    colors = newColors;
+}
+
+QColor UserPimerLineEdit::getColorForInsertedText(int pos) const {
+    QColor resultColor;
+    int primerSize = qMax(0, lastTextSize - (leftEndSize + rightEndSize));
+    if (changeLeftEndColor || (0 <= pos && pos < leftEndSize)) {
+        resultColor = Qt::red;
+    } else if (!changeLeftEndColor && !changeRightEndColor && leftEndSize <= pos && pos <= leftEndSize + primerSize) {
```

Comment:
When you reopen the tab, the backlight breaks. I suggest using `text().size()` instead of `lastTextSize` , and replace `pos <= leftEndSize + primerSize` with `pos < leftEndSize + primerSize` before `Qt::black`.



I also don't know if the current behavior is correct: if you select multiple letters, whether they are the same color or not, and insert some other letters instead, they will be inserted in black. For example:
![image](https://user-images.githubusercontent.com/39720063/150507285-4fe33bfc-6cac-42c3-8ad4-5d3ce6b83148.png)